adapters/ogury/ogury.go

```diff
@@ -0,0 +1,209 @@
+		if err != nil {
+			return nil, append(errors, &errortypes.BadInput{
+				Message: "Error while marshaling Imp.Ext bidder exension",
+			})
+		}
+		request.Imp[i].Ext = ext
+
+		// save adUnitCode
+		request.Imp[i].TagID = imp.ID
+	}
+
+	// currency conversion
+	for i, imp := range request.Imp {
+		// Check if imp comes with bid floor amount defined in a foreign currency
+		if imp.BidFloor > 0 && imp.BidFloorCur != "" && strings.ToUpper(imp.BidFloorCur) != "USD" {
+
+			// Convert to US dollars
+			convertedValue, err := requestInfo.ConvertCurrency(imp.BidFloor, imp.BidFloorCur, "USD")
+			if err != nil {
+				return nil, []error{err}
+			}
+
+			// Update after conversion. All imp elements inside request.Imp are shallow copies
+			// therefore, their non-pointer values are not shared memory and are safe to modify.
+			request.Imp[i].BidFloorCur = "USD"
+			request.Imp[i].BidFloor = convertedValue
+		}
+	}
+
+	requestJSON, err := json.Marshal(request)
+	if err != nil {
+		return nil, []error{err}
+	}
+
+	requestData := &adapters.RequestData{
+		Method:  "POST",
+		Uri:     a.endpoint,
+		Body:    requestJSON,
+		Headers: headers,
+		ImpIDs:  openrtb_ext.GetImpIDs(request.Imp),
+	}
+
+	return []*adapters.RequestData{requestData}, nil
+
+}
+
+func filterValidImps(request *openrtb2.BidRequest) (validImps []openrtb2.Imp) {
+	for _, imp := range request.Imp {
+		var impExt adapters.ExtImpBidder
+		var impExtOgury openrtb_ext.ImpExtOgury
+
+		if err := json.Unmarshal(imp.Ext, &impExt); err != nil {
+			continue
+		}
+		if err := json.Unmarshal(impExt.Bidder, &impExtOgury); err != nil {
+			continue
+		}
+		if impExtOgury.AssetKey != "" && impExtOgury.AdUnitID != "" {
+			validImps = append(validImps, imp)
+		}
+	}
+
+	// if we have imp with assetKey/adUnitId then we want to serve them
+	if len(validImps) > 0 {
+		return validImps
+	}
+
+	// no assetKey/adUnitId imps then we serve everything if publisher.ID exists
+	if request.Site != nil && request.Site.Publisher.ID != "" {
+		return request.Imp
+	}
+
+	// else no valid imp
+	return nil
+}
+
+func setHeaders(request *openrtb2.BidRequest) http.Header {
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	if request.Device != nil {
+		headers.Add("X-Forwarded-For", request.Device.IP)
+		headers.Add("User-Agent", request.Device.UA)
+		headers.Add("Accept-Language", request.Device.Language)
+	}
+	return headers
+
+}
+
+func getMediaTypeForBid(impressions []openrtb2.Imp, bid openrtb2.Bid) (openrtb_ext.BidType, error) {
+	for _, imp := range impressions {
+		if imp.ID == bid.ImpID {
+			switch {
+			case imp.Banner != nil:
+				return openrtb_ext.BidTypeBanner, nil
+			case imp.Video != nil:
+				return openrtb_ext.BidTypeVideo, nil
+			case imp.Native != nil:
+				return openrtb_ext.BidTypeNative, nil
+			}
+		}
+
+	}
+
+	return "", &errortypes.BadServerResponse{
+		Message: fmt.Sprintf("Failed to determine media type of impression \"%s\"", bid.ImpID),
+	}
+}
```

Comment:
This is an anti-pattern. It's preferable for your server to return the bid type, ideally through the `mtype` field.